src/server_manager/web_app/gcp_account.ts

```diff
@@ -63,31 +64,40 @@ export class GcpAccount implements gcp.Account {
   async createServer(projectId: string, name: string, zone: gcp.Zone):
       Promise<server.ManagedServer> {
     const zoneLocator = {projectId, zoneId: zone.id};
-    const instance = await this.createInstance(zoneLocator, name);
-    const id = `${this.id}:${instance.id}`;
-    return new GcpServer(id, instance, this.apiClient);
+    const instanceName = makeGcpInstanceName();
+    const {instanceId, instanceCreation} =
+        await this.createInstance(zoneLocator, name, instanceName);
+    const instanceLocator = {instanceId, ...zoneLocator};
+    const id = `${this.id}:${instanceId}`;
+    return new GcpServer(id, instanceLocator, instanceName, instanceCreation, this.apiClient);
   }
 
   /** @see {@link Account#listServers}. */
   async listServers(projectId: string): Promise<server.ManagedServer[]> {
     const result: GcpServer[] = [];
-    const listZonesResponse = await this.apiClient.listZones(projectId);
-    const listInstancesPromises = [];
-    for (const zone of listZonesResponse.items) {
-      const filter = 'labels.outline=true';
-      const zoneLocator = {projectId, zoneId: zone.name};
-      const listInstancesPromise = this.apiClient.listInstances(zoneLocator, filter);
-      listInstancesPromises.push(listInstancesPromise);
-    }
-    const listInstancesResponses = await Promise.all(listInstancesPromises);
-    for (const response of listInstancesResponses) {
-      const instances = response.items ?? [];
-      instances.forEach((instance) => {
+    const filter = 'labels.outline=true';
+    const [listAllInstancesResponse, listAllStaticIpsResponse] = await Promise.all([
+      await this.apiClient.listAllInstances(projectId, filter),
+      await this.apiClient.listAllStaticIps(projectId)
+    ]);
+    const staticIpMap = listAllStaticIpsResponse?.items ?? {};
+    const ipNames = new Set();
+    Object.values(staticIpMap).forEach(({addresses}) => {
+      addresses?.forEach(ip => ipNames.add(ip.name));
+    });
+    const instanceMap = listAllInstancesResponse?.items ?? {};
+    Object.values(instanceMap).forEach(({instances}) => {
+      instances?.forEach(instance => {
+        const {zoneId} = gcp_api.parseZoneUrl(instance.zone);
+        const locator = {projectId, zoneId, instanceId: instance.id};
+        // Repair missing IP address.  This could happen if the server manager was
+        // closed during server creation.
+        const instanceCreation = ipNames.has(instance.name) ?
+            Promise.resolve() : this.promoteEphemeralIp(locator);
```

Comment:
We are doing this logic in two different places. Imagine there were more actions. It's hard to ensure they are doing the same thing. Let's remove this duplication.

This logic can live where we create the GcpServer. Either in the constructor or in a factory function.

Reply:
OK, I've moved `promoteEphemeralIp` into `GcpServer`, calling it in the constructor when needed.  I've also added a distinction between `instanceCreation` and `instanceReadiness` (i.e instance and static IP are all set up).